Document Efficient solver configuration for nonlinear models (#7165)
geodynamics/aspect: doc/sphinx/user/run-aspect/run-faster/efficent_stokes_solver_configuration_for_nonlinear_models.md

# Efficient Stokes solver configuration for nonlinear models

## Background

One of the major challenges in geodynamic modeling is that the constitutive equations can be highly nonlinear. 
A common cause is the nonlinear rheological behavior of mantle rocks, which can make viscosity depend on a power of the strain-rate invariant, as in dislocation creep and plastic yielding.

ASPECT already provides nonlinear solver schemes to address this numerical challenge. Rather than diving into their scientific meaning or technical implementation, this cookbook focuses on a more specific and practical question that every user must deal with: How should the existing nonlinear solver be configured to achieve robust and efficient convergence in a simulation?

> The primary consideration is, of course, choosing Stokes solver parameters that ensure convergence. However, even when convergence is achieved, the solver configuration may still have the problem of being overly expensive. For example, the convergence tolerance may be set unnecessarily low, or the chosen solver scheme may be too rigid for the problem.

Reviewer: Saying that a "tolerance is low" is misunderstandable because tolerance is something like accuracy. I prefer to say "tolerance is tight".

> Ideally, the solver configuration should be just sufficient to achieve convergence within the desired numerical tolerance while minimizing computational cost.

Reviewer: From a mathematical perspective, you never actually *achieve* convergence. Convergence is an asymptotic process: You can get closer and closer to the exact solution, but you will never actually reach it. Perhaps say something like this instead?

```suggestion
Ideally, the solver configuration should be just sufficient to compute a solution that is sufficiently accurate (i.e., within the desired numerical tolerance) while minimizing computational cost.
```

> ## A short introduction to ASPECT's nonlinear solving scheme

Reviewer: ```suggestion
## A short introduction to ASPECT's nonlinear solvor scheme
```

Author: Changed this and related places.

When a problem is nonlinear, each time step requires solving a nonlinear Stokes system. 

> Here we show an example with the set up in the `crust_model_2d` cookbook. 

Reviewer: Can you reference the section in the manual here that discusses this cookbook?

> The nonlinear solver first defines the required solution accuracy through the parameter `Nonlinear solver tolerance`. It then achieves this target through a sequence of nonlinear iterations, with each iteration bringing the solution closer to the desired accuracy.

Reviewer: ```suggestion
The nonlinear solver first defines the required solution accuracy through the parameter `Nonlinear solver tolerance`. It then achieves this target through a sequence of nonlinear iterations, with each iteration bringing the current numerical approximation closer to the exact solution, stopping when the desired accuracy has been reached.
```

| Setup dof systems               |        44 |     0.156s |         0% |
| Setup initial conditions        |         2 |   0.00554s |         0% |
| Setup matrices                  |        43 |     0.242s |      0.13% |
| Solve Stokes system             |      1242 |       169s |        91% |
+---------------------------------+-----------+------------+------------+
```

In general, we want a robust solver configuration that is capable of handling the difficult initial stage of the simulation.
However, such a configuration can also be unnecessarily expensive because the same strict linear solver settings are retained throughout the simulation, even after the nonlinear problem has become much easier to solve. This is the case in the previous example, which uses a large number of expensive Stokes solver iterations throughout the simulation.

Moreover, this type of costly solver configuration often results from a typical workflow. When the first simulation attempts fail because of solver convergence issues, it is tempting to simply adopt a more robust solver configuration without a second thought on the transient nature of the non-convergence.

> ---

Reviewer: This creates a horizontal line, right? Why do you need this?

Author: removed this and other instances.

```

In this example, one could indeed enlarge the "Number of cheap Stokes solver steps" to converge the nonlinear iteration solely with cheap Stokes solver, for instance, in the current example a number of 2000 would converge most nonlinear iterations.
However, this is not the purpose of this strategy, which is to reduce computation cost.
There are no substantial differences in the wall-clock time with a number of 200 vs 2000.

---

## Strategy B: Introducing a linear solver failure strategy to continue with the nonlinear solver

Strategy A demonstrates a philosophy for handling the challenging initial time steps by prioritizing any configuration that allows the nonlinear solver to converge, while choosing linear solver parameters that provide good performance during the later, easier stages of the simulation.

However, Strategy A still requires the linear solver to be sufficiently robust to achieve the required linear solver tolerance.
Otherwise, the simulation terminates when a linear solve fails.

This philosophy can be extended further by allowing the linear solver to fail occasionally during nonlinear iterations.

> Instead of terminating the simulation, the nonlinear solver continues using the solution from the failed linear solve and proceeds to the next nonlinear iteration, where another linear solve is attempted.

Reviewer: ```suggestion
Instead of terminating the simulation, the nonlinear solver continues using the solution from the failed linear solve and proceeds to the next nonlinear iteration, where another linear solve is attempted. The underlying philosophy here is that in a nonlinear scheme, it is necessary to solve each linear system *exactly*; all we care about is that the computed solution of the linear system helps us move *closer* to the solution of the nonlinear problem. That might be the case even if we can't reach the desired linear solver tolerance.
```

> ASPECT now provides the `Linear solver failure strategy` option to support this behavior:

Reviewer: ```suggestion
ASPECT provides the `Linear solver failure strategy` option to support this behavior:
```
(The use of "now" indicates something recent, but people will still read this section in ten years.)
